### PR TITLE
feat: add NeuralDQNPolicy with experience replay and target network (closes #20)

### LINKED ISSUE #20: Policy: Neural DQN with Experience Replay and Target Network
## Summary

Add a **DQN policy** (`NeuralDQNPolicy`) that replaces the discretised tabular Q-table in `EpsilonGreedyPolicy` / `MCTSPolicy` with a small MLP function approximator trained via experience replay and a frozen target network.

## Motivation

`EpsilonGreedyPolicy` and `MCTSPolicy` both discretise the 21-dimensional observation into 3ⁿ buckets (default n_bins=3 → ~10 billion states at full dimensionality — in practice most states are never visited). This causes:

- Extremely slow value propagation across state space
- No generalisation between similar but unseen states
- State aliasing (two different states map to same bin tuple)

A neural Q-function `Q(s, a; θ)` generalises across the continuous observation space and converges orders of magnitude faster.

## Proposed Implementation

Add `NeuralDQNPolicy` to `policies.py`:

```python
class NeuralDQNPolicy(BasePolicy):
    """
    DQN over the 9-action discrete action set.
    Online network: Q(s, a; θ_online)
    Target network: Q(s, a; θ_target)  (synced every target_update_freq steps)
    Replay buffer: circular buffer of (s, a, r, s', done)
    """
```

### Architecture

```
Input: obs (21 + n_lidar_rays dims)
Hidden: [64, 64]  (ReLU)
Output: Q-values for each of 9 discrete actions
```

All weights managed in pure numpy (consistent with `NeuralNetPolicy`). Gradient computed analytically via backprop.

### Algorithm sketch

```
Initialise:
  online_net, target_net  (same small MLP as NeuralNetPolicy)
  replay_buffer           (deque, maxlen=replay_buffer_size)
  epsilon = epsilon_start → epsilon_end via linear schedule

Each step (update(obs, action, reward, next_obs, done)):
  1. Push (s, a_idx, r, s', done) to replay_buffer
  2. if len(buffer) < min_replay_size: return
  3. Sample minibatch of size batch_size
  4. Compute targets: y = r + γ * max_a Q(s'; θ_target)  [zero if done]
  5. Compute loss: MSE(Q(s, a; θ_online), y)
  6. Backprop through online_net, update θ_online (Adam / SGD)
  7. Every target_update_freq steps: θ_target ← θ_online

Action selection (epsilon-greedy over Q-values):
  if rand() < epsilon: random action from _DISCRETE_ACTIONS
  else: argmax_a Q(obs; θ_online)
```

### Config additions

```yaml
policy_type: neural_dqn
policy_params:
  hidden_sizes: [64, 64]
  replay_buffer_size: 10000
  batch_size: 64
  min_replay_size: 500       # steps before first update
  target_update_freq: 200    # steps between target syncs
  learning_rate: 0.001
  epsilon_start: 1.0
  epsilon_end: 0.05
  epsilon_decay_steps: 5000  # linear decay over this many steps
  gamma: 0.99
```

## Files to change

| File | Change |
|------|--------|
| `policies.py` | Add `NeuralDQNPolicy`, `ReplayBuffer` helper (~200 lines) |
| `main.py` `_make_policy()` | Register `"neural_dqn"` case |
| `config/training_params.yaml` | Document new policy type + params |
| `tests/test_policies.py` | Unit test: Q-values converge on a fixed MDP |

## Key Design Decisions

- **Pure numpy backprop**: consistent with `NeuralNetPolicy`; avoids torch/tf dependency. Implement a minimal autograd-free SGD pass over the two-layer MLP.
- **9 discrete actions**: reuse `_DISCRETE_ACTIONS` already defined at `policies.py:40`. No change to action space.
- **Online learning**: `update()` is called inside the episode loop (same as `QTablePolicy`). Works with the existing `on_episode_end()` hook for epsilon scheduling.
- **Save format**: serialize weights (same nested-list YAML as `NeuralNetPolicy`) + hyperparams.

## Acceptance Criteria

- [ ] `NeuralDQNPolicy` inherits `BasePolicy`, implements `__call__`, `update`, `on_episode_end`, `to_cfg`, `save`
- [ ] Replay buffer is circular with configurable max size
- [ ] Target network syncs every `target_update_freq` gradient steps
- [ ] Epsilon decays linearly from `epsilon_start` to `epsilon_end` over `epsilon_decay_steps`
- [ ] `policy_type: neural_dqn` works end-to-end in `main.py`
- [ ] Unit test: policy achieves non-trivial Q-value convergence on a 2-state synthetic MDP

## References

- Mnih et al. (2015). *Human-level control through deep reinforcement learning.* Nature 518, 529–533.

### PR DESCRIPTION
**Pull request by espenhk**
Implements DQN over the 9-action discrete action set in pure numpy:
- ReplayBuffer: circular buffer (collections.deque) of (s, a, r, s', done)
- NeuralDQNPolicy: MLP Q-function with online + frozen target network,
  Adam optimiser, linear epsilon schedule, and YAML serialisation
- main.py: registers "neural_dqn" in _make_policy() with full cfg passthrough
- config/training_params.yaml: documents all neural_dqn hyperparameters
- tests/test_neural_dqn_policy.py: 16 unit tests including bandit convergence

https://claude.ai/code/session_01Newa1Qvg4P732JmhtqzQqj

### CODE REVIEW

**Comment by espenhk** [comment]
Closes #20

File: policies.py [hunk 1]
```diff
@@ -807,3 +808,302 @@ def save(self, path: str) -> None:
...
+
+    def sample(self, batch_size: int) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+        idxs  = np.random.choice(len(self._buf), size=batch_size, replace=False)
+        batch = [self._buf[i] for i in idxs]
+        obs_b  = np.stack([t[0] for t in batch]).astype(np.float32)
+        act_b  = np.array([t[1] for t in batch], dtype=np.int32)
+        rew_b  = np.array([t[2] for t in batch], dtype=np.float32)
+        next_b = np.stack([t[3] for t in batch]).astype(np.float32)
+        done_b = np.array([t[4] for t in batch], dtype=np.float32)
+        return obs_b, act_b, rew_b, next_b, done_b
```
[the remark this answers is not in the document]
Comment (espenhk) [reply policies.py:837]: @copilot apply changes based on [this feedback](https://github.com/espenhk/tmnf-ai/pull/26#discussion_r3067784582)

File: policies.py [hunk 2]
```diff
@@ -807,3 +808,302 @@ def save(self, path: str) -> None:
...
+        )
+        if "online_weights" in cfg:
+            obj._online["weights"] = [np.array(w, dtype=np.float32) for w in cfg["online_weights"]]
+            obj._online["biases"]  = [np.array(b, dtype=np.float32) for b in cfg["online_biases"]]
+            obj._target["weights"] = [np.array(w, dtype=np.float32) for w in cfg["target_weights"]]
+            obj._target["biases"]  = [np.array(b, dtype=np.float32) for b in cfg["target_biases"]]
+            obj._eps         = float(cfg.get("epsilon",      obj._eps_end))
+            obj._total_steps = int(cfg.get("total_steps",   0))
+            obj._grad_steps  = int(cfg.get("grad_steps",    0))
+            # Re-init Adam moments for the restored weights
```
[the remark this answers is not in the document]
Comment (espenhk) [reply policies.py:959]: @copilot apply changes based on [this feedback](https://github.com/espenhk/tmnf-ai/pull/26#discussion_r3067784591)

File: tests/test_neural_dqn_policy.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    def test_bandit_convergence(self):
+        N = BASE_OBS_DIM
+        state   = np.zeros(N, dtype=np.float32)
+        BEST    = 7          # accel + straight
+        GOOD_R  =  1.0
+        BAD_R   = -0.1
+
+        policy = NeuralDQNPolicy(
+            hidden_sizes        = [32, 32],
+            replay_buffer_size  = 5000,
```
[the remark this answers is not in the document]
Comment (espenhk) [reply tests/test_neural_dqn_policy.py:210]: @copilot apply changes based on [this feedback](https://github.com/espenhk/tmnf-ai/pull/26#discussion_r3067784593)
